### PR TITLE
Bug 1340683 - stylo: Implement the :-moz-any pseudo-class

### PR DESCRIPTION
**Pull request by mbrubeck**
Adds support for the non-standard [:-moz-any](https://developer.mozilla.org/en-US/docs/Web/CSS/:any) selector.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix https://bugzilla.mozilla.org/show_bug.cgi?id=1340683
- [x] These changes do not require tests because they are gecko-only

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15966)

### CODE REVIEW

**Comment by mbrubeck** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=04585b67f0646e335d87e48584f4abe7fd6d9a64

File: components/selectors/parser.rs [hunk 1]
```diff
@@ -120,7 +120,7 @@ pub trait Parser {
     }
 }
 
-#[derive(PartialEq, Clone, Debug)]
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
```
Comment (emilio) [inline components/selectors/parser.rs]: Is this change needed just due to the bounds in `SelectorImpl`s definition?

If so, I believe we can remove them since we switched to conditional compilation in `style` and should have a concrete type now.

Also, in pseudo-classes it doesn't really matter I think, since we only store `PseudoElement`s in `HashMap`s.

**Comment by mbrubeck** [comment]
This still fails the `layout/style/test/test_any_dynamic.html` mochitest; it doesn't handle dynamic updates correctly.  I need to update `affects_siblings` and other selector methods to take `:-moz-any` into account.

File: components/style/gecko/wrapper.rs [hunk 2]
```diff
@@ -661,7 +661,8 @@ impl<'le> ::selectors::Element for GeckoElement<'le> {
             NonTSPseudoClass::MozTableBorderNonzero |
             NonTSPseudoClass::MozBrowserFrame => unsafe {
                 Gecko_MatchesElement(pseudo_class.to_gecko_pseudoclasstype().unwrap(), self.0)
-            }
+            },
+            NonTSPseudoClass::MozAny(ref selectors) => matches(&selectors.0, self, None),
```
Comment (emilio) [inline components/style/gecko/wrapper.rs]: This doesn't pass the `ElementFlags`, nor the `StyleRelations` here. I believe it should be ok, because we unconditionally insert the appropriate flag for all the custom non-tree-structural pseudo-classes, and because of the limit on only-descendant combinators.

Can you verify that something like http://searchfox.org/mozilla-central/source/layout/reftests/css-selectors/state-dependent-in-any.html (perhaps tweaking manually with `:hover` or `:active`, or something we actually implement instead of `:valid`) works?

**Comment by emilio** [comment]
Oh, I had already started reviewing before you commented :)

Upon reflection, I believe the dynamic changes will not pass until https://bugzilla.mozilla.org/show_bug.cgi?id=1338982 is fixed (I'm going to fix it in https://bugzilla.mozilla.org/show_bug.cgi?id=1345950, but I wanted #15890 to land first)

File: components/style/gecko/selector_parser.rs [hunk 3]
```diff
@@ -264,6 +274,24 @@ impl<'a> ::selectors::Parser for SelectorParser<'a> {
         }
     }
 
+    fn parse_non_ts_functional_pseudo_class(&self, name: Cow<str>, parser: &mut CssParser)
+                                            -> Result<NonTSPseudoClass, ()> {
+        match_ignore_ascii_case! { &name,
+            "-moz-any" => {
+                // https://developer.mozilla.org/en-US/docs/Web/CSS/:any
+                // The selectors may not contain pseudo-elements and the only combinator allowed is
```
Comment (bzbarsky) [inline components/style/gecko/selector_parser.rs]: The devmo docs don't seem to match the code.  The code has:


    // Check that none of the selectors in the list have combinators or
    // pseudo-elements. 

and proceeds to do just that.  See http://searchfox.org/mozilla-central/rev/ca7015fa45b30b29176fbaa70ba0a36fe9263c38/layout/style/nsCSSParser.cpp#6511-6518

**Review by bzbarsky — CHANGES_REQUESTED** [review]

**Comment by bzbarsky** [comment]
Note that we also have tests for this at http://searchfox.org/mozilla-central/rev/ca7015fa45b30b29176fbaa70ba0a36fe9263c38/layout/style/test/test_selectors.html#1184-1209 and the second test there tests that descendant combinators don't parse...

But it looks like lack of support for other things earlier in the test (e.g. :placeholder-shown) causes it to throw exceptions that prevent it from even reaching this part.  :(  

Maybe we should change should_serialize_to to more gracefully handle cases when the selector is not supported at all.  :(  @upsuper

File: components/style/gecko/selector_parser.rs (hunk 3 above)
Comment (mbrubeck) [inline components/style/gecko/selector_parser.rs]: Fixed to match the Gecko code.

**Comment by mbrubeck** [comment]
> I need to update affects_siblings and other selector methods to take :-moz-any into account.

Done. I think this is ready for re-review now.

File: components/style/gecko/wrapper.rs (hunk 2 above)
Comment (mbrubeck) [inline components/style/gecko/wrapper.rs]: > Can you verify that something like http://searchfox.org/mozilla-central/source/layout/reftests/css-selectors/state-dependent-in-any.html (perhaps tweaking manually with :hover or :active, or something we actually implement instead of :valid) works?

Selectors like `:-moz-any(:hover)` do not seem to update correctly on state changes; I'm not certain why, since as you said `AFFECTED_BY_STATE` is always set.

Also, things like `:-moz-any(:first-child)` are valid and should set `HAS_EDGE_CHILD_SELECTOR` and `AFFECTED_BY_CHILD_INDEX`, so I think we do need to pass the relations and flags here.

Comment (mbrubeck) [inline components/style/gecko/wrapper.rs]: Ah, part of the problem is that `pseudo_class_state_flag` was not implemented correctly for MozAny.

**Comment by mbrubeck** [comment]
Updated to correctly calculate `ElementStyleFlags`, `ElementState`, and `StyleRelations` for selectors like `:-moz-any(:hover)`.

File: components/style/gecko/wrapper.rs (hunk 2 above)
Comment (mbrubeck) [inline components/style/gecko/wrapper.rs]: Fixed.

**Comment by mbrubeck** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=bfef2302480ed879637ea957ba3e047631d4f5f6

**Comment by mbrubeck** [comment]
Rebased.

File: components/style/gecko/selector_parser.rs [hunk 4]
```diff
@@ -207,6 +245,10 @@ impl NonTSPseudoClass {
                 match *self {
                     $(NonTSPseudoClass::$name => flag!($state),)*
                     $(NonTSPseudoClass::$s_name(..) => flag!($s_state),)*
+                    NonTSPseudoClass::MozAny(ref selectors) => {
+                        selectors.iter().fold(ElementState::empty(), |state, s|
```
Comment (emilio) [inline components/style/gecko/selector_parser.rs]: nit: since the closure wraps to the next line, let's just wrap it with braces.

File: components/style/gecko/selector_parser.rs [hunk 5]
```diff
@@ -293,6 +336,11 @@ impl<'a> ::selectors::Parser for SelectorParser<'a> {
                         let name = String::from(parser.expect_ident_or_string()?).into_boxed_str();
                         NonTSPseudoClass::$s_name(name)
                     }, )*
+                    "-moz-any" => {
+                        let selectors = parser.parse_comma_separated(|input|
+                            ComplexSelector::parse(self, input))?;
```
Comment (emilio) [inline components/style/gecko/selector_parser.rs]: same here.

File: components/style/restyle_hints.rs [hunk 6]
```diff
@@ -347,6 +350,13 @@ impl<'a, E> Element for ElementWrapper<'a, E>
     }
 }
 
+/// Returns the union of any `ElementState` flags for components of a `ComplexSelector`
+pub fn complex_selector_to_state(sel: &ComplexSelector<SelectorImpl>) -> ElementState {
+    sel.compound_selector.iter().fold(ElementState::empty(), |state, s|
+        state | selector_to_state(s)
```
Comment (emilio) [inline components/style/restyle_hints.rs]: and here.

**Review by emilio — APPROVED** [review]
Looks good. We've talked over IRC and we'll make sure there are bugs on file for the remaining failures in `layout/style/test/test_any_dynamic.html`, and `layout/style/test/test_bug635286.html`.

I'm pretty sure they'll be fixed by:
 * https://bugzilla.mozilla.org/show_bug.cgi?id=1338982
 * https://bugzilla.mozilla.org/show_bug.cgi?id=1345950

But better getting them on file.

**Comment by emilio** [comment]
Whoops.

@bors-servo r-

Please address the review nits, then r=me

File: components/style/gecko/selector_parser.rs (hunk 4 above)
Comment (mbrubeck) [inline components/style/gecko/selector_parser.rs]: Done.

File: components/style/gecko/selector_parser.rs (hunk 5 above)
Comment (mbrubeck) [inline components/style/gecko/selector_parser.rs]: Done.

File: components/style/restyle_hints.rs (hunk 6 above)
Comment (mbrubeck) [inline components/style/restyle_hints.rs]: Done.
